fax/test_util.py

```diff
@@ -188,7 +195,216 @@ def func(u):
     def equality_constraints(u):
         return np.linalg.norm(u) - 1
 
-    optimal_solution = -(1./np.linalg.norm(v))*v
+    optimal_solution = -(1. / np.linalg.norm(v)) * v
     optimal_value = np.dot(optimal_solution, v)
 
     return func, equality_constraints, optimal_solution, optimal_value
+
+
+def get_list(rows):
+    param_list = []
+    skipped = []
+    for row in rows:
+        row_text = row.lstrip()
+        if not row_text:
+            continue
+
+        if row_text.startswith("!"):
+            skipped.append(row_text)
+            continue
+
+        if {">=", "<=", "<", ">"}.intersection(row_text):
+            raise NotImplementedError("no inequalities")
+
+        if row_text[0].isupper() and row.replace(" ", "").isalpha():
+            assert row_text.startswith("End")
+            return param_list, skipped
+        else:
+            param_list.append(row_text)
+    raise ValueError
+
+
+def get_struct(rows):
+    struct = {}
+    skipped = []
+    for row in rows:
+        if not row:
+            continue
+
+        if row[0] == '!' or row[0] == '#':
+            skipped.append(row)
+            continue
+
+        row_text = row.lstrip()
+        if row_text == "End Model":
+            continue
+
+        if row_text[0].isupper():
+            struct[row_text], skipped_ = get_struct(rows)
+            skipped.extend(skipped_)
+        else:
+            params, skipped_ = get_list(itertools.chain([row], rows))
+            skipped.extend(skipped_)
+            return params, skipped
+    return struct, skipped
+
+
+def text_to_code(variable, equation, closure):
+    cost_function = equation.replace("^", "**")
+    seq = []
+    for a in cost_function.split("]"):
+        if "[" not in a:
+            seq.append(a)
+        else:
+            rest, num = a.split("[")
+            b = f"{rest}[{int(num) - 1}"
+            seq.append(b)
+    cost_function = "]".join(seq)
+    scope = ", ".join(k for k in closure.keys() if not k.startswith("__"))
+    cost_function_ = f"{variable} = lambda {scope}: {cost_function}"
+    return cost_function_
+
+
+def apm_to_python(text: Text) -> Union[Text, None]:
```

Comment:
Probably ok for now, but perhaps at some point would be good to clean up test_util.py and separate things out into different files